assets/lambda/ImageOptimization.ts

```diff
@@ -0,0 +1,160 @@
+import { Readable } from 'stream'
+import https from 'https'
+
+const sourceBucket = process.env.S3_SOURCE_BUCKET ?? undefined
+
+/**
+	* The imageConfig object is stringified and passthrough the lambda as an environment variable
+	*/
+const _loadImageConfig = (): ImageConfig => {
+	let imageConfig: ImageConfig = {}
+	const imageConfigString = process.env.NEXT_IMAGE_CONFIG
+	try {
+		imageConfig = JSON.parse(imageConfigString || '{}') as ImageConfig
+	} catch (err) {
+		console.error('Error parsing image config: ', { err, imageConfigString })
+		throw err
+	}
+
+	return imageConfig
+}
+
+const _getRequest = (url: NextUrlWithParsedQuery): Promise<Buffer> => {
+	return new Promise((resolve, reject) => {
+		const req = https.get(url, res => {
+			const chunks: Buffer[] = []
+			res.on('data', chunk => chunks.push(chunk))
+			res.once('end', () => resolve(Buffer.concat(chunks)))
+			res.once('error', err => {
+				reject(err)
+			})
+		})
+		req.once('error', err => {
+			reject(err)
+		})
+	})
+}
+
+// Handle fetching of S3 object before optimization happens in nextjs.
+const requestHandler =
+	(bucketName: string) =>
+	async (req: IncomingMessage, res: ServerResponse, url?: NextUrlWithParsedQuery): Promise<void> => {
+		if (!url) {
+			throw new Error('URL is missing from request.')
+		}
+
+		let response: any
+		let data: Buffer
+		// External url, try to fetch image
+		if ( !!url.href?.startsWith('http')) {
+			try {
+				data = await _getRequest(url)
+			} catch (err) {
+				console.error('Failed to get image', err)
+				res.statusCode = 400
+				res.end()
+				return
+			}
+
+		} else {
+				// S3 expects keys without leading `/`
+			const trimmedKey = url.href.startsWith('/') ? url.href.substring(1) : url.href
+
+			const client = new S3Client({})
+			response = await client.send(new GetObjectCommand({ Bucket: bucketName, Key: trimmedKey }))
+			if (!response.Body) {
+				throw new Error(`Could not fetch image ${trimmedKey} from bucket.`)
+			}
+			const stream = response.Body as Readable
+
+			data = await new Promise<Buffer>((resolve, reject) => {
+				const chunks: Buffer[] = []
+				stream.on('data', (chunk) => chunks.push(chunk))
+				stream.once('end', () => resolve(Buffer.concat(chunks)))
+				stream.once('error', reject)
+			})
+		}
+
+		res.statusCode = 200
+
+		if (response.ContentType) {
+			res.setHeader('Content-Type', response.ContentType)
+		}
+
+		if (response.CacheControl) {
+			res.setHeader('Cache-Control', response.CacheControl)
+		}
+
+		res.write(data)
+		res.end()
+	}
+
+// Make header keys lowercase to ensure integrity.
+const normalizeHeaders = (headers: Record<string, any>) =>
+	Object.entries(headers).reduce((acc, [key, value]) => ({ ...acc, [key.toLowerCase()]: value }), {} as Record<string, string>)
+
+// Nextjs expects domains to be a URL object.
+	const imageConfig = _loadImageConfig()
+// @TODO: Allow passing params as env vars.
+const nextConfig = {
```

Comment:
do we have a type for this?
maybe it should use the same config as the lambda handler? that contains the next.config.js resolved config btw...

Reply:
I'll type it as `NextConfigComplete`